Nftables rule scale recipe

SirPhuttel opened the pull request:
Add a recipe suited for testing nftables rule traversal performance (or the
other way round, a given rule's impact on packet processing). It takes a rule
and the number of times to add it to the ruleset as parameter and generates
results using FlowMeasurementGenerator. Sample instantiation looks like this:
```
r = NftablesRuleScaleRecipe(perf_tests=["udp_stream"],
                            offload_combinations=[],
			    rule="fib saddr . iif oif missing drop",
			    scale=1000)
Controller().run(recipe_instance)
```

The first commit in this PR adds SimpleNetnsRouterRecipe which enables testing
the kernel's forwarding path without the need for three hosts and a dedicated
link between two of them. It simply creates a netns on the second host and
inserts the netns-facing VETH interface as second endpoint instead of the
second host's physical NIC.

NftablesRuleScaleRecipe added by the second commit makes use of the netns
router by adding rules to the second host's netfilter forward hook. This should
prevent the test setup from disturbing control traffic between the agents.

jtluka commented on lnst/Recipes/ENRT/SimpleNetnsRouterRecipe.py:
```diff
@@ -0,0 +1,96 @@
+
+class SimpleNetnsRouterRecipe(SimpleNetworkRecipe):
+    """
+    This recipe implements Enrt testing for a simple network scenario that
```
Please add some notes what this recipe is about.

jtluka commented on lnst/Recipes/ENRT/SimpleNetnsRouterRecipe.py:
```diff
@@ -0,0 +1,96 @@
+        host2.run("sysctl -w net.ipv4.ip_forward=1")
+        host2.run("sysctl -w net.ipv6.conf.all.forwarding=1")
+
+        ipv4_addr = interface_addresses(self.params.net_ipv4)
```
Somewhat confusing - could we change the previous ipv4_addr to ipv4_ns_addr or similar?

jtluka commented on lnst/Recipes/ENRT/SimpleNetnsRouterRecipe.py:
```diff
@@ -0,0 +1,96 @@
+        host2.run("sysctl -w net.ipv6.conf.all.forwarding=1")
+
+        ipv4_addr = interface_addresses(self.params.net_ipv4)
+        next(ipv4_addr)
```
I'm not quite sure this is necessary? Why we need to skip one address?a

SirPhuttel replied:
This is because I need the second address generated from `self.params.net_ipv4`. It is the host1-facing external IP of host2 which I use to create a route to the netns-network on host1.
As said, maybe an extra getter (or one I didn't discover yet) will simplify things here.

SirPhuttel replied:
I hope it's less chaotic now, noticed I can extract the IPs from `self.params.net_ipv4` directly via an index.

jtluka reviewed (changes requested):
Left few questions and comments, overall looks ok, but check my comments.

SirPhuttel commented on lnst/Recipes/ENRT/NftablesRuleScaleRecipe.py:
```diff
@@ -0,0 +1,50 @@
+from lnst.Recipes.ENRT.MeasurementGenerators.FlowMeasurementGenerator import FlowMeasurementGenerator
+from lnst.Common.Parameters import StrParam, IntParam, BoolParam
+
+class NftablesRuleScaleRecipe(SimpleNetnsRouterRecipe, NftablesMixin, FlowMeasurementGenerator):
```
Oh, thanks!

SirPhuttel commented on lnst/Recipes/ENRT/NftablesRuleScaleRecipe.py:
```diff
@@ -0,0 +1,50 @@
+
+    @property
+    def firewall_rulesets_generator(self):
+        rule = self.params.get('rule', None)
```
Oh, right. I took those calls from sample code and didn't check what that second parameter was for. A classic case of cargo cult! /o\

SirPhuttel commented:
Changes since v1:
- Improve docstrings of added classes
- Eliminate the need for temporary address variables in SimpleNetnsRouterRecipe by fetching the addresses from NetworkParams via an index
- Break some long lines
- NftablesRuleScaleRecipe does not need to explicitly inherit FlowMeasurementGenerator
- Drop pointless second parameter from self.params.get() calls
- Simplify ruleset construction in NftablesRuleScaleRecipe
- Add NftablesRuleScaleRecipe to docs/

olichtne commented on lnst/Recipes/ENRT/SimpleNetnsRouterRecipe.py:
```diff
@@ -0,0 +1,95 @@
+        host2.ns.np0.up_and_wait()
+
+        for gw in (self.params.netns_ipv4[1], self.params.netns_ipv6[1]):
+            host2.ns.np0._ipr_wrapper("route", "add", dst="default",
```
not entirely sure i like this call of a "private" wrapper method from the recipe, this should probably added as a new public Device class method `add_route`... but i think for now let's keep it as is...

n0-1 replied:
Guess I could also just call `host2.run('ip -net host2ns route add ...')`, not sure if this improves things or not though.

olichtne replied:
yeah, that's a bit better, it at least ensures that the recipe won't stop working if we decide to change the private method somehow...
i'd do this though:
```
host2.ns.run("ip route add...")
```
to run the command in the ns directly

SirPhuttel replied:
Oh, sorry Ondrej! I forgot about this open issue (and don't ask me what I assumed it was about when I marked this one as resolved. /o\

olichtne commented on lnst/Recipes/ENRT/SimpleNetnsRouterRecipe.py:
```diff
@@ -0,0 +1,95 @@
+                                      gateway=f"{gw}")
+
+        host2.run("sysctl -w net.ipv4.ip_forward=1")
+        host2.run("sysctl -w net.ipv6.conf.all.forwarding=1")
```
unless these are default config values, i think these need the reverse action added to the `test_wide_deconfiguration` method, otherwise they may affect other tests running in sequence after this one.

we used to have a tester api for `sysctl` that used to take care of this automatically but we don't have this right now...

olichtne replied:
this applies to other calls of `host.run('sysctl...')` as well...

n0-1 replied:
Ah, thanks! I missed that. Do I have to kill the netns as well? If so, will `host2.ns = None` do? (Please pardon my ignorance.)

olichtne replied:
namespaces are cleaned up automatically so just the sysctls will do, thanks!

olichtne commented:
Looks good to me, i think this can be merged unless @jtluka has some other comments.

@SirPhuttel did you try to run this in beaker or just locally?

Maybe we can add one test run using the internal LNST infrastructure if you can provide some configuration example to just have a double check that everything works as expected. I can schedule that no problem if you provide me with the configuration.

SirPhuttel commented:
Hi @olichtne,

> @SirPhuttel did you try to run this in beaker or just locally?

Just locally, using some Fedora VMs.

> Maybe we can add one test run using the internal LNST infrastructure if you can provide some configuration example to just have a double check that everything works as expected. I can schedule that no problem if you provide me with the configuration.

That would be great! My goto is:
```
ctl = Controller()
r = NftablesRuleScaleRecipe(perf_tests=["udp_stream"],
			    offload_combinations=[],
			    rule="fib saddr . iif oif missing drop",
			    scale=10)
ctl.run(r)
```
Cheers, Phil
